Sources/LiveKit/Core/Engine.swift

```diff
@@ -211,23 +211,52 @@ class Engine: MulticastDelegate<EngineDelegate> {
             return
         }
 
+        guard publisher.iceConnectionState != .checking else {
+            return
+        }
+
         hasPublished = true
         publisher.negotiate()
     }
 
+    internal func publisherDataChannel(for reliability: Reliability) -> RTCDataChannel? {
+        reliability == .reliable ? dcReliablePub : dcLossyPub
+    }
+
     internal func send(userPacket: Livekit_UserPacket,
-                       reliability: DataPublishReliability = .reliable) -> Promise<Void> {
+                       reliability: Reliability = .reliable) -> Promise<Void> {
+
+        func ensurePublisherConnected () -> Promise<Void> {
+
+            guard let publisher = publisher else {
+                return Promise(EngineError.invalidState("publisher is nil"))
+            }
+
+            guard subscriberPrimary,
+                  !publisher.isIceConnected,
+                  publisherDataChannel(for: reliability)?.readyState != .open else {
+                // aleady connected and ready, no-op
+                return Promise(())
+            }
+
+            publisherShouldNegotiate()
+
+            return waitForIceConnect(transport: publisher).then {
```

Comment:
FWIW, you could just do `waitForPublisherDataChannelOpen` directly. When dc.readyState changes, it should be good to go.